File: Heap/python/heap.py
class Heap:

    def __init__(self, comparator=lambda a, b: a > b): 

        self.array = []
        self.comparator = comparator
        
    def _swap(self, x, y):    
        self.array[x], self.array[y] = self.array[y], self.array[x]

    def isEmpty(self):
        return len(self.array) == 0
    
    def toArray(self):
        return this.array.copy()

    def size(self):
        return len(self.array)

    def getHeap(self):
        return self.array.copy() 
    
    def clear(self):        
        self.array = []

    def _heapify(self, position):

        if position >= self.size():
            return
        
        rightPos = lambda x : (x * 2) + 2
        leftPos = lambda x : (x * 2) + 1

        max = position
        right = rightPos(position)
        left = leftPos(position)

        if(left < self.size() and self.comparator(self.array[max], self.array[left]) == False):
            max = left

        if(right < self.size() and self.comparator(self.array[max], self.array[right]) == False):
            max = right

        if max != position:
            self._swap(position, max)
            self._heapify(max)

    def insert(self, element):

        if (len(self.array) > 0 and type(element) != type(self.array[0])) or element == None:
            return

        self.array.append(element)
        self.upHeap(self.size() - 1)

    def upHeap(self, position):

        if position > 0:

            parentPos = lambda x : (x - 1) / 2

            parent = parentPos(position)

            if self.comparator(self.array[position], self.array[parent]) == True:

                self._swap(position, parent)
                self._heapify(position)
                self.upHeap(parent)

    def buildHeap(self, array):
        
        if array == None or len(array) == 0:
            return
        
        self.clear()
        map(self.insert, array)

    def extractRootElement(self):

        if self.size() == 0:
            return None
        
        else:
            
            root = self.array[0]

            if self.size() > 1:
                self.array[0] = self.array.pop(self.size() - 1)
                self._heapify(0)

            else:
                self.clear()

            return root

    def rootElement(self):
        return self.array[0] if self.size() > 0 else None
    
    def heapSort(self, array):

        out = []

        if array == None:
            return out
        
        comp, l = self.comparator, self.array

        self.comparator = lambda a, b: a < b
        self.clear()

        map(self.insert, array)
        
        [out.append(self.extractRootElement()) for i in xrange(self.size())]

        self.clear()
        self.comparator, self.array = comp, l

        return out


if __name__ == "__main__":

    h = Heap()
    assert h.size() == 0
    assert h.isEmpty() == True

    h.buildHeap([82, 6, 99, 12, 34, 64, 58, 1])
    assert h.size() == 8
    assert h.isEmpty() == False
    assert h.array == [99, 34, 82, 6, 12, 64, 58, 1]
    assert h.extractRootElement() == 99
    assert h.rootElement() == 82

    h.clear()
    assert h.size() == 0
    assert h.isEmpty() == True
    assert h.heapSort([82, 6, 99, 12, 34, 64, 58, 1]) == sorted([82, 6, 99, 12, 34, 64, 58, 1])

File: Heap/javascript/heap.js
const BinaryHeap = () => [];

const comparator = (a,b) => a > b ? true : false;

const size = heap => heap.length;

const _parent = index => parseInt((index - 1) / 2);

const left = index => (index * 2) + 1;

const right = index => (index * 2) + 2;

const isEmpty = heap => (heap.length === 0);

const toArray = heap => heap;

const rootElement = heap => !isEmpty(heap) ? heap[0] : undefined;

const swap = (heap, index_1, index_2) => {
   const newHeap = [...heap];
   const item_1 = newHeap[index_1];
   newHeap[index_1] = newHeap[index_2];
   newHeap[index_2] = item_1;
   return newHeap;   
};

const heapify = (heap, index) => {
   let newHeap = [...heap];
   if(index >= size(heap)){
      return newHeap;
   }

   let max = index;

   if(left(index) < size(heap) && !comparator(newHeap[max], newHeap[left(index)])){
      max = left(index);
   }

   if(right(index) < size(heap) && !comparator(newHeap[max], newHeap[right(index)])){
      max = right(index);
   }

   if(max !== index){
      newHeap = swap(newHeap, index, max);
      newHeap = heapify(newHeap, max);
   }

   return newHeap;
};


const upHeap = (heap, index) => {
   if (index < 0){
      return [...heap];
   }

   let newHeap = [...heap];
   const parent_index = _parent(index);

   if(comparator(heap[index], heap[parent_index])){
      newHeap = swap(newHeap, index, parent_index);
      newHeap = heapify(newHeap, index);
      newHeap = upHeap(newHeap, parent_index);
   }

   return newHeap;
};

const insert = (heap, element) => {
   let newHeap = [...heap, element];
   newHeap = upHeap(newHeap, size(newHeap) - 1);
   return newHeap;
};

const buildHeap = (array) => array.reduce((heap, elem) => {
    return insert(heap, elem); 
 }, BinaryHeap());

 const extractRootElement = heap => {
   if(isEmpty(heap)){
      return heap;
   }

   let newHeap = [...heap];
   newHeap[0] = newHeap.pop();
   newHeap = heapify(newHeap, 0);
   return newHeap;
};

const heapSort = array => (array !== null) ? toArray(buildHeap(array)) : [];
